src/Client/discovery.ts

```diff
@@ -64,74 +64,75 @@ export async function discoverEndpoint(
   }
 
   throw new Error(`Failed to discover after ${discoverAttempts} attempts.`);
-}
+};
 
-function inAllowedStates(member: MemberInfo): boolean {
-  switch (member.state) {
-    case VNodeState.SHUTDOWN:
-      return false;
+const allowedStates = new Set([
+  VNodeState.FOLLOWER,
+  VNodeState.LEADER,
+  VNodeState.READONLYREPLICA,
+  VNodeState.PREREADONLYREPLICA,
+  VNodeState.READONLYLEADERLESS,
+]);
+
+export const isInAllowedState = (member: MemberInfo): boolean =>
+  member.isAlive && allowedStates.has(member.state);
+
+// getPreferedStates, higher index is better
+const getPreferedStates = (preference: NodePreference) => {
+  switch (preference) {
+    case LEADER:
+      return [VNodeState.LEADER];
+    case FOLLOWER:
+      return [VNodeState.FOLLOWER];
+    case READ_ONLY_REPLICA:
+      return [
+        VNodeState.READONLYLEADERLESS,
+        VNodeState.PREREADONLYREPLICA,
+        VNodeState.READONLYREPLICA,
+      ];
     default:
-      return true;
+      return [];
   }
-}
+};
+
+type CompareFn = (a: MemberInfo, b: MemberInfo) => number;
+const compareByPreference = (preference: NodePreference): CompareFn => {
+  const preferedStates = getPreferedStates(preference);
+  return (a, b) =>
+    preferedStates.indexOf(b.state) - preferedStates.indexOf(a.state);
+};
+const shuffle: CompareFn = (a, b) => Math.random() - 0.5;
```

Comment:
**CONCERN:** That may not be an issue here, but `Math.random` is a pseudo-random number generator. For the real randomness [`Crypto.getRandomValues](https://nodejs.org/api/webcrypto.html#webcrypto_crypto_getrandomvalues_typedarray)` should be used.